### PR TITLE
[APM] Only show relevant service legends

### PR DESCRIPTION
**Pull request by sqren**
This PR hides service legends that are not currently visible in the timeline because the the trace is "zoomed in".

This should also fix an issue highlighted by @simitt where legend colors change even when services stay in the same order.

@jasonrhodes @ogupte

Closes elastic/kibana#24741
Closes elastic/kibana#24818

### CODE REVIEW

File: x-pack/plugins/apm/typings/waterfall.ts [hunk 1]
```diff
@@ -7,7 +7,4 @@
 import { Span } from './Span';
 import { Transaction } from './Transaction';
 
-export interface WaterfallResponse {
-  services: string[];
-  hits: Array<Transaction | Span>;
-}
+export type WaterfallResponse = Array<Transaction | Span>;
```
Comment (jasonrhodes) [inline x-pack/plugins/apm/typings/waterfall.ts]: oh wow I only now realize how much more I like this :clap:

Comment (sqren) [reply x-pack/plugins/apm/typings/waterfall.ts]: Yes! This change was so easy to make. I just updated this interface, and then the compiler guided me through everywhere I needed to update something. I find myself writing code without ever running it in the browser - until I push.

**Comment by jasonrhodes** [comment]
@sqren I just tested pushing to your branch with a manually checked out branch, and it worked great. I added the tiny services label to the legend to complete that other ticket, here's what that looks like (cc @formgeist)

screen shot 2018-10-30 at 12 15 40 pm

File: x-pack/plugins/apm/public/components/app/TransactionDetails/Transaction/WaterfallContainer/Waterfall/waterfall_helpers/waterfall_helpers.ts [hunk 2]
```diff
@@ -124,16 +132,20 @@ export function getWaterfallItems(
...
+function getTraceRoot(childrenByParentId: IWaterfallGroup) {
   const item = first(childrenByParentId.root);
   if (item && item.docType === 'transaction') {
     return item.transaction;
   }
-};
+}
+
+function getServices(items: IWaterfallItem[]) {
+  const serviceNames = items.map(item => item.serviceName);
```
Comment (ogupte) [inline x-pack/plugins/apm/public/components/app/TransactionDetails/Transaction/WaterfallContainer/Waterfall/waterfall_helpers/waterfall_helpers.ts]: could just use the lodash `map` here `return uniq(map(items, 'serviceName'))`

Comment (sqren) [reply x-pack/plugins/apm/public/components/app/TransactionDetails/Transaction/WaterfallContainer/Waterfall/waterfall_helpers/waterfall_helpers.ts]: Call me old fashioned but for equivalent methods (map, filter, reduce, find etc.) I always prefer native. Especially after advent of ES6 arrow syntax I'm no longer as envious of lodash' iteratee shorthand.

Comment (sqren) [reply x-pack/plugins/apm/public/components/app/TransactionDetails/Transaction/WaterfallContainer/Waterfall/waterfall_helpers/waterfall_helpers.ts]: Any reason I should reconsider?

**Comment by sqren** [comment]
Awesome, thanks @jasonrhodes. Merging.

File: x-pack/plugins/apm/public/components/app/TransactionDetails/Transaction/WaterfallContainer/Waterfall/waterfall_helpers/waterfall_helpers.ts (hunk 2 above)
Comment (ogupte) [reply x-pack/plugins/apm/public/components/app/TransactionDetails/Transaction/WaterfallContainer/Waterfall/waterfall_helpers/waterfall_helpers.ts]: in most browsers, lodash map is much more performant than native map

Comment (ogupte) [reply x-pack/plugins/apm/public/components/app/TransactionDetails/Transaction/WaterfallContainer/Waterfall/waterfall_helpers/waterfall_helpers.ts]: this was true for past versions of browser, but it seems that recent version are catching up in the benchmarks (https://jsperf.com/native-map-vs-lodash-map/10)

Comment (sqren) [reply x-pack/plugins/apm/public/components/app/TransactionDetails/Transaction/WaterfallContainer/Waterfall/waterfall_helpers/waterfall_helpers.ts]: Yeah browsers improve - Lodash does not. I know... that's a controversial statement :p
But Kibana are still stuck on Lodash 3 while users are free to run evergreen browsers.

Regardless, I doubt any speed difference is noticeable for the end user in this case :)

Comment (ogupte) [reply x-pack/plugins/apm/public/components/app/TransactionDetails/Transaction/WaterfallContainer/Waterfall/waterfall_helpers/waterfall_helpers.ts]: too true 💯

**Pull request merged**